sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HadoopFileSystemOptions.java

```diff
@@ -39,11 +45,37 @@
   void setHdfsConfiguration(List<Configuration> value);
 
   /** A {@link DefaultValueFactory} which locates a Hadoop {@link Configuration}. */
-  class ConfigurationLocator implements DefaultValueFactory<Configuration> {
+  class ConfigurationLocator implements DefaultValueFactory<List<Configuration>> {
+    private static final Logger LOG = LoggerFactory.getLogger(ConfigurationLocator.class);
     @Override
-    public Configuration create(PipelineOptions options) {
-      // TODO: Find default configuration to use
-      return null;
+    public List<Configuration> create(PipelineOptions options) {
+      // Find default configuration when HADOOP_CONF_DIR or YARN_CONF_DIR is set.
+      Configuration conf = new Configuration(false);
+      List<String> hadoopEnvList = Lists.newArrayList("HADOOP_CONF_DIR", "YARN_CONF_DIR");
+      for (String env : hadoopEnvList) {
+        String hadoopConfPath = System.getenv(env);
+        if (!Strings.isNullOrEmpty(hadoopConfPath) && new File(hadoopConfPath).exists()) {
+
+          // We just need to load both core-site.xml and hdfs-site.xml to determine the
+          // default fs path and the hdfs configuration
+          if (new File(hadoopConfPath + "/core-site.xml").exists()) {
+            conf.addResource(new Path(hadoopConfPath + "/core-site.xml"));
+
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Adding " + hadoopConfPath + "/core-site.xml to hadoop configuration");
+            }
+          }
+
+          if (new File(hadoopConfPath + "/hdfs-site.xml").exists()) {
+            conf.addResource(new Path(hadoopConfPath + "/hdfs-site.xml"));
+
+            if (LOG.isDebugEnabled()) {
+              LOG.debug("Adding " + hadoopConfPath + "/hdfs-site.xml to hadoop configuration");
+            }
+          }
+        }
+      }
+      return Lists.<Configuration>newArrayList(conf);
```

Comment:
We shouldn't be returning a configuration if we didn't load one from one of the paths.